src/services/storage/storage.service.ts

```diff
@@ -0,0 +1,183 @@
+ * Retreives array from the extension storage
+ * @param key entry's key
+ * @param factory function that creates default array if it is not found
+ * @returns array from the extension storage or an array with default values
+ */
+export async function getArray<T extends object>(key: string, factory: () => T[]): Promise<T[]> {
+  const array = await getEntry<T>(key)
+
+  return Array.isArray(array) ? array : factory()
+}
+
+/**
+ * Abstracts access to the extension storage.
+ * Provides default values for non existing entries
+ */
+export class Storage {
+  private listeners: Record<string, Array<(entry: unknown) => void>> = {}
+
+  static readonly networkKey = 'network'
+  static readonly networkListKey = 'network-list'
+  static readonly swarmKey = 'swarm'
+
+  constructor() {
+    chrome.storage.onChanged.addListener(this.onChangeListener.bind(this))
+  }
+
+  public getNetworkList(): Promise<Network[]> {
+    return getArray<Network>(Storage.networkListKey, networkListFactory)
+  }
+
+  public async addNetworkToList(network: Network): Promise<void> {
+    const networks = await this.getNetworkList()
+
+    if (!networks.find(({ label }) => network.label === label)) {
+      networks.push(network)
+      await setArray<Network>(Storage.networkListKey, networks)
+    }
+  }
+
+  public async updateNetworkInList(label: string, network: Network): Promise<void> {
+    const networks = await this.getNetworkList()
+    const index = networks.findIndex((net) => net.label === label)
+
+    if (index >= 0) {
+      networks[index] = network
+      await setEntry<Network[]>(Storage.networkListKey, networks)
+    }
+  }
+
+  public async deleteNetworkFromList(network: Network): Promise<void> {
+    const networks = await this.getNetworkList()
+    const index = networks.findIndex(({ label }) => network.label === label)
+
+    if (index >= 0) {
+      networks.splice(index, 1)
+      await setEntry<Network[]>(Storage.networkListKey, networks)
+    }
+  }
+
+  public getNetwork(): Promise<Network> {
+    return getObject<Network>(Storage.networkKey, networkFactory)
+  }
+
+  public setNetwork(network: Network): Promise<void> {
+    return setObject<Network>(Storage.networkKey, network)
+  }
+
+  public getSwarm(): Promise<Swarm> {
+    return getObject<Swarm>(Storage.swarmKey, swarmFactory)
+  }
+
+  public setSwarm(swarm: Swarm): Promise<void> {
+    return setObject<Swarm>(Storage.swarmKey, swarm)
+  }
+
+  public onNetworkChange(listener: (network: Network) => void) {
+    this.setListener(Storage.networkKey, listener)
+  }
+
+  public onSwarmChange(listener: (swarm: Swarm) => void) {
+    this.setListener(Storage.swarmKey, listener)
+  }
+
+  public removeListener(listener: (entry: unknown) => void) {
+    Object.values(this.listeners).forEach((listeners) => {
+      removeAllValues(listeners, listener)
+    })
+  }
+
+  /**
+   * Migrates existing data to match the current version
+   * @returns promise
+   */
+  public migrate(): Promise<void> {
```

Comment:
can you explain this function please?

Reply:
I've described it in one of the previous comments.